source/contributed/tutorialMonkeyPatching.md

```diff
@@ -0,0 +1,203 @@
+# Monkey Patching Tutorial for Screeps
+WARNING: Monkey patching is often considered bad practice and there are those that will tell you that you should NEVER do it, and there are good reasons for this. Monkey patching, while very powerful and convenient (especially in Screeps), is a dangerous practice that should not be taken lightly.
+
+Monkey patching is the practice of modifying the way that existing prototype functions operate.
+
+We will provide examples that will demonstrate a few of the infinite uses for monkey patching each example will try to build on the previous one, so you'll want to read through all of it to make sure you understand it fully.
+
+
+## Introduction - Adding functions to prototypes
+
+To understand monkey patching it is important to understand that functions are/can be stored in variables just like numbers, objects, arrays, and strings. Creep functions like `.moveTo`, `.harvest`, or `.withdraw` are all stored in variables belonging to the Creep prototype object. These functions can be accessed like so: `Creep.prototype.moveTo`, `Creep.prototype.harvest`, and `Creep.prototype.withdraw`. Any values that belong to the Creep prototype will be accessible from any Creep object.  This is why you are able to call `creep.moveTo` on all of your creeps. 
+
+Because these functions are variables in an object just like any other, you can add new ones: 
+```javascript
+Creep.prototype.sayHello = function() { 
+    // In prototype functions, 'this' usually has the value of the object calling the function. in this case that is whatever creep you are calling 'sayHello' on.
+    this.say("Hello!"); 
+};
+```
+after this code, you can do `creep.sayHello();` on any of your creeps and they will say "Hello!"
+
+You can overwrite existing functions:
+```javascript
+Creep.prototype.suicide = function() {
+    this.say("NO WAY MAN");
+};
+```
+The above code overwrites the normal `creep.suicide` function so that instead of suiciding, the creep will voice his disagreement with the command.
+
+But.. what if you really need the creep to suicide? You no longer have the normal suicide function because you overwrote it with our own!!! (in game this is not permanent. It would regain its original value on a global reset, this also means you can execute these code examples outside of your main loop, they dont need to be executed every tick.
+
+# EXAMPLE #1 - upgradeController - "Praise GCL"
+Let's try another one but a little differently. We'll modify the `.upgradeController` function so that the creep says "Praise GCL" whenever it upgrades the controller. First, unlike what we did with `.suicide`, we are going to store the original function so we don't lose it.
+### Store the original function
+```javascript
+// Store the original function
+Creep.prototype.storedUpgradeController = Creep.prototype.upgradeController;
+// or
+let storedUpgradeController = Creep.prototype.upgradeController;
```

Comment:
Should probably encourage the use of `const`, especially on something like this.